### PR TITLE
Add support for using Annoy as an external similarity index

### PR DESCRIPTION
**Pull request by droudy**
(Related to #617, #527, and #51.)

Implementation + tests + tutorial

### CODE REVIEW

**Comment by tmylk** [comment]
Thanks for the PR to @droudy and @karlhigley!

File: docs/notebooks/annoytutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+   "metadata": {},
+   "source": [
+    "# Similarity Queries using Annoy Tutorial"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "This tutorial is about using the Annoy(Approximate Nearest Neighbors Oh Yeah) library for similarity queries in gensim"
```
Comment (piskvorky) [inline docs/notebooks/annoytutorial.ipynb]: Add hyperlink to Annoy here.

File: docs/notebooks/annoytutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    "neighbors = annoy_index.most_similar(vector, 5)\n",
+    "for neighbor in neighbors:\n",
+    "    print neighbor"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A similarity query using Annoy is significantly faster then using the traditional brute force method\n",
```
Comment (piskvorky) [inline docs/notebooks/annoytutorial.ipynb]: then => than

File: docs/notebooks/annoytutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    ">**Note**: Initialization time for the annoy indexer was not included in the times. The optimal knn algorithm for you to use will depend on how many queries you need to make and the size of the corpus. If you are making very few similarity queries, the time taken to initialize the annoy indexer will be longer then the time it would take the brute force method to retrieve results. If you are making many queries however, the time it takes to initialize the annoy indexer will be made up for by the incredibly fast retrieval times for queries once the indexer has been initialized"
```
Comment (piskvorky) [inline docs/notebooks/annoytutorial.ipynb]: then => than

File: docs/notebooks/annoytutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    "## What is Annoy?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "[Annoy](https://github.com/spotify/annoy \"Link to annoy repo\") is an open source library to search for points in space that are close to a given query point. It also creates large read-only file-based data structures that are mmapped into memory so that many processes may share the same data. For our purpose, it is used to find similarity between words or documents in a vector space. [See the tutorial on similarity queries for more information on them](https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/Similarity_Queries.ipynb).\n",
+    "\n",
+    "[There are benefits](https://github.com/spotify/annoy#background) of using annoy over the pre-existing method of making similarity queries through brute force in gensim."
```
Comment (piskvorky) [inline docs/notebooks/annoytutorial.ipynb]: Gensim indexes are even easier to memory-map, so this is no benefit over gensim (comment incorrect).

**Comment by piskvorky** [comment]
Nice functionality! Thank you. I added comment for fixing a few minor typos.

@tmylk I'd still add some conclusions / CTA section at the end. The ending reads a little weird.

File: docs/notebooks/annoytutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    "model = gensim.models.Word2Vec.load('/tmp/leemodel')\n",
+    "print model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Using the SimilarityIndex class\n",
+    "An instance of `SimilarityIndex` needs to be created in order to use Annoy in gensim. The `SimilarityIndex` class is located in `gensim/similarities/index`\n",
```
Comment (piskvorky) [inline docs/notebooks/annoytutorial.ipynb]: If it's meant to be a filesystem path, use extension too (and perhaps even hyperlink).

File: gensim/similarities/index.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+try:
+    from annoy import AnnoyIndex
+except ImportError:
+    raise ImportError("Annoy has not been installed, if you wish to use the annoy indexer, please run `pip install annoy`")
+
+
+class SimilarityIndex(object):
+
+    @classmethod
+    def build_from_word2vec(cls, model, num_trees):
```
Comment (piskvorky) [inline gensim/similarities/index.py]: I feel this is too specialized. Better API: pass in a 2d numpy matrix. @gojomo thoughts?

File: gensim/similarities/index.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    def most_similar(self, vector, num_neighbors):
+        """Find the top-N most similar items"""
+
+        ids, distances = self.index.get_nns_by_vector(
+            vector, num_neighbors, include_distances=True)
+
+        result = []
+        for i in range(len(ids)):
+            label = self.labels[ids[i]]
+            similarity = 1 - distances[i] / 2
```
Comment (piskvorky) [inline gensim/similarities/index.py]: Has this been verified? I remember there was some confusing around the actual formula used in gensim, https://github.com/spotify/annoy/issues/149.

File: gensim/similarities/index.py (hunk 6 above)
Comment (tmylk) [inline gensim/similarities/index.py]: I like that this is beginner-friendly. No need ot know what `syn0norm` is.

Comment (piskvorky) [inline gensim/similarities/index.py]: We can make specialized wrappers for specific models, but the core API functionality should definitely be an arbitrary matrix/sequence of vectors (like the existing Similarity interfaces).

File: gensim/similarities/index.py (hunk 7 above)
Comment (karlhigley) [inline gensim/similarities/index.py]: This appears to be correct, according to the Annoy README and spotify/annoy#162.

**Pull request merged**